### ISSUE TITLE
Python 3.8 does not have SSL module on macOS agent

### ISSUE THREAD

**Issue by junyuan-z** [actions/virtual-environments#1745]
**Description**  
We are running into a problem related to the SSL module on macOS agent.

Previously we download files like this:

```
with requests.get(url, allow_redirects=True) as response, open(handler, "wb") as output:
        output.write(response.content)
```

the URL could be HTTP or HTTPS.

starting from October 5th we are running into issues related to SSL module

```
requests.exceptions.SSLError: HTTPSConnectionPool(host='buildreqs.blob.core.windows.net', port=443): Max retries exceeded with url: /java/macOS/jdk10u.tar.gz (Caused by SSLError("Can't connect to HTTPS URL because the SSL module is not available."))
```

I think this may have something to do with the macOS base image.

**Area for Triage**:  Python

**Question, Bug, or Feature?**:  Bug

**Virtual environments affected**  
- [x] macOS 10.15
- [ ] Ubuntu 16.04 LTS
- [ ] Ubuntu 18.04 LTS
- [ ] Ubuntu 20.04 LTS
- [ ] Windows Server 2016 R2
- [ ] Windows Server 2019

**Expected behaviour**  

the previously working code should still work.

**Actual behaviour**  

broken due to some external factors.

**Repro steps**  
A description with steps to reproduce the issue. If you have a public example or repo to share,
please provide the link.

write a simple python script that tries to download something from HTTPS resource like this:

```
with requests.get(url, allow_redirects=True) as response, open(handler, "wb") as output:
        output.write(response.content)
```

run this in the macOS agent and you will see the error

**Comment by junyzheng** [comment]
Figured out the problem. There are 2 OpenSSL installed on macOS agent: 1.0.2 and 1.1.1. The 1.0.2 OpenSSL is linked to `/usr/local/bin`. However, because it is an older formula comparing to 1.1.1, it will be removed if users run `brew install` first. I am proposing to clean up the installed OpenSSL library on the macOS agent.

**Comment by andy-mishechkin** [comment]
Hello, @junyzheng 
Thank you for your note.
We'll investigate if we can remove OpenSSL 1.0.2 from image, if it will not impact to other software in image.

**Comment by dmitry-shibanov** [comment]
Hello @junyzheng  , we have investigated possibility to get rid of openssl 1.0.2 on images and set 1.1.1 as default one. We have found that this change breaks Python versions installed on image. We are investigating on possible solutions

**Comment by maxim-lobanov** [comment]
Thank you for reporting the issue and your proposal. We think that it definitely makes sense to get rid of openssl 1.0.2 (at least, because it is not supported after the end of 2019) and set `openssl 1.1.1` as default.
Unfortunately, it is not so simple to do because:
1. Python on image is built with openssl 1.0.2 and it doesn't work properly with openssl 1.1.1
1. Ruby compatibility issue with openssl 1.0.2
1. Some AzureDevOps tasks (related to certificates) don't support openssl 1.1.1
1. Changing default openssl can impact customers who build own applications via GA / ADO

So we need to resolve 1 and 2 from our side and do proper communication on 3 and 4.
We will keep this issue opened and will plan this work this quarter

**Comment by eregon** [comment]
I noticed that `brew --prefix openssl` was once pointing to OpenSSL 1.1.1, and then more recently to 1.0.2:
https://github.com/ruby/setup-ruby/issues/97#issuecomment-715422992
As a result a older build of Ruby ended up building against 1.1 and a newer build of Ruby ended up building against 1.0.2.

That's rather surprising. Also, upstream Homebrew has `openssl` pointing to `openssl@1.1`:
https://github.com/Homebrew/homebrew-core/blob/master/Aliases/openssl
So this is some kind of incompatibility.

`brew --prefix openssl` is used by [`ruby-install`](https://github.com/postmodern/ruby-install/blob/c339bf6ca481ec6fac7fd18ec537f46bed0d1ec5/share/ruby-install/ruby/functions.sh#L24) but also other software.

https://github.com/actions/virtual-environments/pull/1808 seems the right way.

**Comment by eregon** [comment]
> Python on image is built with openssl 1.0.2 and it doesn't work properly with openssl 1.1.1

@maxim-lobanov Could we rebuild Python against openssl 1.1.1? Is there any issue with that? (it would be surprising if recent Python doesn't support OpenSSL 1.1)

> Ruby compatibility issue with openssl 1.0.2

That's only for Ruby <= 2.3., which is not provided in the hostedtoolcache AFAIK.
https://github.com/ruby/setup-ruby/issues/97#issuecomment-715422992 uses https://github.com/ruby/ruby-builder/blob/2f5047a2b2249fa9bfb342c0fee0baba93ffb9ee/.github/workflows/build.yml#L155-L156 which uses `ruby-build`, which builds its own OpenSSL 1.0.2 for such old versions.
From that comment, OpenSSL 1.0.2 was (at the time) impossible to install on Ubuntu 16.04, and that's probably still the case, and even more so on 20.04.
In short, I think there is no issue regarding Ruby for GitHub Actions.

If we cannot plainly remove OpenSSL 1.0.2 on macOS soon, I would suggest to at least let `brew --prefix openssl` and `/usr/local/opt/openssl` point to OpenSSL 1.1.
That is, remove this code: https://github.com/actions/virtual-environments/blob/0d46520ccf08c3780f879b324d81639975be3fcd/images/macos/provision/core/openssl.sh#L18-L22

**Comment by maxim-lobanov** [comment]
@eregon , thank you for your input and thoughts about Ruby. 
We are planning to rebuild Python soon and switch openssl after that.

**Comment by maxim-lobanov** [comment]
Just an update: Python was rebuilt with openssl 1.1 and this change was deployed everywhere.
We have created announcement for deprecation openssl 1.0.2 and setup openssl 1.1 as default on image: https://github.com/actions/virtual-environments/issues/2089
It will happen in couple of weeks.

**Comment by kdeldycke** [comment]
Now that #2089 has been addressed, I guess we can close this issue right?

**Comment by maxim-lobanov** [comment]
Yes, new image has been deployed.
It contains only openssl 1.1 so we can close this issue.
Please let me know if you have any concerns.
